test scenarios for sync to genesis after restart when historical sync is working on blocks from era 0 created pre-1.5

alsrdn opened the pull request:
Before 1.5.0 there wasn't an immediate switch block committed at genesis.
Because historical sync relies on sync leaps to get the validators for a particular era, we encounter a special case when syncing the blocks of era 0 if they were created before 1.5.0 because sync leap will not be able to include a switch block that has the validators for era 0.
Since the auction delay is at least 1, we can generally rely on the fact that the validator set for era 1 and era 0 are the same.

Test if a node that has synced back to some block in era 0 can continue syncing to genesis after it was restarted (and lost its validator matrix).

Addresses: https://github.com/casper-network/casper-node/issues/4077

alsrdn committed:
nctl: test scenario for sync to genesis after restart when sync in era 0

Before 1.5.0 there wasn't an immediate switch block committed at
genesis. Because historical sync relies on sync leaps to get the
validators for a particular era, we encounter a special case
when syncing the blocks of era 0 if they were created before 1.5.0
because sync leap will not be able to include a switch block that
has the validators for era 0.
Since the auction delay is at least 1, we can generally rely on
the fact that the validator set for era 1 and era 0 are the same.

Add a test to check if a node that has synced back to some block in
era 0 can continue syncing to genesis after it was restarted (and
lost its validator matrix).

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
tests/integration: check if nodes can sync back with 1 block eras

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

fizyk20 commented on node/src/reactor/main_reactor/tests.rs:
```diff
@@ -225,6 +240,21 @@ fn has_completed_era(era_id: EraId) -> impl Fn(&Nodes) -> bool {
     }
 }
 
+fn lowest_available_block_height_on_node(height: u64, node_id: NodeId) -> impl Fn(&Nodes) -> bool {
```
This function's name is a bit confusing... but I have no ideas for a good one :grimacing: Something like `create_lowest_available_block_height_check`, as it creates a closure that checks something, maybe?

alsrdn replied:
Yeah, indeed it returns a predicate for checking if the specified height is the lowest available. Maybe if I rename it to `is_lowest_available_block_height_on_node` and also add a comment it would be in line with the other checks like `is_in_era` and `has_completed_era`. WDYT?

fizyk20 replied:
Ah, I missed the other two similar functions.
Hm, so to bring this back in line, how about `has_lowest_available_block_below_height`, to stress the property of the height we're looking for? Or perhaps `node_has_lowest_available_block_below_height`, to also stress that it only looks at a single node?

alsrdn replied:
Yup, sounds good to me, I'll make the change.

alsrdn replied:
renamed to `node_has_lowest_available_block_at_or_below_height` in https://github.com/casper-network/casper-node/pull/4149/commits/7c1334f55e7430f11baba52f630ea966be6b929a

fizyk20 reviewed (approved):
Looks good :+1: Only one small nitpick from me.

alsrdn committed:
tests/integration: rename predicate for checking lowest available block

Rename `lowest_available_block_height_on_node` to
`node_has_lowest_available_block_at_or_below_height` in order to be
consistent with the other predicates.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

cspramit commented:
needs a 2nd review @EdHastingsCasperLabs

EdHastingsCasperAssociation commented on node/src/reactor/main_reactor/tests.rs:
```diff
@@ -341,6 +376,113 @@ async fn run_network() {
     .await;
 }
 
+fn highest_finalized_block_hash(
```
`highest_complete_block_hash`

alsrdn committed:
tests/integration/main_reactor: rename helper function

Rename `highest_finalized_block_hash` to `highest_complete_block_hash`
in main reactor integration tests.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
bors r+

alsrdn commented:
bors r+
